File: src/types/aiLifecycle.ts
/**
 * Phase 13.20: Autonomous AI Application Lifecycle Platform (AAILP) Types.
 * Frontend TypeScript Interfaces & Enums.
 */

export type AgentLifecycleState =
  | 'DRAFT'
  | 'DEVELOPMENT'
  | 'TESTING'
  | 'SECURITY_REVIEW'
  | 'APPROVED'
  | 'DEPLOYED'
  | 'DEPRECATED'
  | 'RETIRED';

export type AgentCategory =
  | 'EXTRACTION'
  | 'RESEARCH'
  | 'CUSTOMER_SUPPORT'
  | 'AUTOMATION'
  | 'COMPLIANCE'
  | 'FINANCIAL_AUDIT'
  | 'LEGAL_ANALYSIS';

export type DeploymentEnvironment = 'STAGING' | 'CANARY' | 'PRODUCTION';
export type DeploymentStrategy = 'DIRECT' | 'BLUE_GREEN' | 'CANARY';
export type DeploymentStatus = 'BUILDING' | 'TESTING' | 'STAGING' | 'CANARY' | 'PRODUCTION' | 'ROLLBACK' | 'FAILED';
export type ApprovalStage = 'DEVELOPER_SUBMIT' | 'SECURITY_REVIEW' | 'BUSINESS_APPROVAL' | 'COMPLIANCE_SIGNOFF' | 'FINAL_RELEASE';
export type ApprovalDecision = 'PENDING' | 'APPROVED' | 'REJECTED';

export interface AgentApplication {
  agent_id: string;
  tenant_id: string;
  organization_id: string;
  workspace_id: string;
  name: string;
  slug: string;
  category: AgentCategory;
  owner_id: string;
  owner_email: string;
  description: string;
  lifecycle_state: AgentLifecycleState;
  current_version: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface AgentVersion {
  version_id: string;
  agent_id: string;
  version_tag: string;
  model_family: string;
  system_prompt: string;
  tools: string[];
  connectors: string[];
  parameters: Record<string, any>;
  changelog: string;
  accuracy_score: number;
  cost_per_execution_usd: number;
  created_at: string;
}

export interface AgentTestResult {
  test_id: string;
  agent_id: string;
  version_tag: string;
  functional_pass: boolean;
  grounding_score: number;
  accuracy_score: number;
  hallucination_rate_pct: number;
  security_checks_passed: boolean;
  latency_p95_ms: number;
  cost_estimated_usd: number;
  status: string;
  tested_at: string;
}

export interface SecurityVulnerability {
  vuln_id: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  category: string;
  description: string;
  recommendation: string;
}

export interface AgentSecurityScan {
  scan_id: string;
  agent_id: string;
  version_tag: string;
  security_score: number;
  risk_level: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  prompt_injection_resistance_pct: number;
  pii_leakage_detected: boolean;
  excessive_permissions: boolean;
  vulnerabilities: SecurityVulnerability[];
  scanned_at: string;
}

export interface AgentApproval {
  approval_id: string;
  agent_id: string;
  version_tag: string;
  stage: ApprovalStage;
  decision: ApprovalDecision;
  approver_id?: string | null;
  approver_email?: string | null;
  comments: string;
  timestamp: string;
}

export interface AgentDeployment {
  deployment_id: string;
  agent_id: string;
  version_tag: string;
  environment: DeploymentEnvironment;
  strategy: DeploymentStrategy;
  traffic_weight_pct: number;
  status: DeploymentStatus;
  distributed_cluster_id: string;
  active_instances_count: number;
  deployed_at: string;
}

export interface AgentDependency {
  dep_id: string;
  agent_id: string;
  dependency_type: string;
  target_resource_id: string;
  target_resource_name: string;
  is_breaking_change: boolean;
}

export interface AgentTemplate {
  template_id: string;
  name: string;
  category: AgentCategory;
  description: string;
  recommended_model: string;
  default_tools: string[];
  default_prompt: string;
}

export interface AgentMarketplaceListing {
  listing_id: string;
  agent_id: string;
  title: string;
  publisher_name: string;
  category: AgentCategory;
  description: string;
  version: string;
  rating: number;
  install_count: number;
  certified_secure: boolean;
  price_monthly_usd: number;
  tags: string[];
}

export interface AgentAnalytics {
  agent_id: string;
  name: string;
  total_executions: number;
  success_rate_pct: number;
  automation_roi_usd: number;
  developer_hours_saved: number;
  adoption_score: number;
  avg_latency_ms: number;
}

export interface AgentRetirementPlan {
  retirement_id: string;
  agent_id: string;
  target_migration_agent_id?: string | null;
  deprecation_notice: string;
  sunset_date: string;
  traffic_redirect_pct: number;
  archived: boolean;
}

export interface LifecycleOverview {
  total_managed_agents: number;
  deployed_in_production: number;
  in_review_or_testing: number;
  deprecated_or_retired: number;
  mean_security_score: number;
  total_automation_roi_usd: number;
  timestamp: string;
}

File: app/api/v1/endpoints/ai_lifecycle.py
"""
Phase 13.20: Autonomous AI Application Lifecycle Platform (AAILP) API Endpoints.
Mounted under /api/v1/ai-lifecycle.
"""

from typing import Dict, List, Optional, Any
from fastapi import APIRouter, HTTPException, Query, Body
from app.platform_ai_lifecycle.runtime.lifecycle_master_orchestrator import LifecycleMasterOrchestrator
from app.platform_ai_lifecycle.models.schemas import (
    AgentApplication,
    AgentCategory,
    AgentLifecycleState,
    AgentVersion,
    AgentTestResult,
    AgentSecurityScan,
    AgentApproval,
    ApprovalStage,
    ApprovalDecision,
    AgentDeployment,
    DeploymentEnvironment,
    DeploymentStrategy,
    AgentDependency,
    AgentTemplate,
    AgentMarketplaceListing,
    AgentAnalytics,
    AgentRetirementPlan,
    LifecycleOverview,
)

router = APIRouter()
orchestrator = LifecycleMasterOrchestrator()


# 1. Executive Control Plane & Overview
@router.get("/overview", response_model=LifecycleOverview, tags=["AI Lifecycle Executive Control"])
async def get_lifecycle_overview():
    """Returns platform-wide AI application lifecycle statistics & ROI."""
    return orchestrator.get_overview()


@router.post("/full-release", tags=["AI Lifecycle Executive Control"])
async def full_lifecycle_release(payload: Dict[str, Any] = Body(...)):
    """Runs end-to-end automated DevSecOps release pipeline for an AI application."""
    required = ["tenant_id", "organization_id", "workspace_id", "name", "slug", "owner_id", "owner_email", "system_prompt"]
    if not all(k in payload for k in required):
        raise HTTPException(status_code=400, detail="Missing required release fields")

    cat_str = payload.get("category", "AUTOMATION")
    category = AgentCategory(cat_str) if cat_str in AgentCategory.__members__ else AgentCategory.AUTOMATION

    return orchestrator.full_lifecycle_release(
        tenant_id=payload["tenant_id"],
        organization_id=payload["organization_id"],
        workspace_id=payload["workspace_id"],
        name=payload["name"],
        slug=payload["slug"],
        category=category,
        owner_id=payload["owner_id"],
        owner_email=payload["owner_email"],
        system_prompt=payload["system_prompt"],
        tools=payload.get("tools", []),
        connectors=payload.get("connectors", []),
    )


# 2. Agent Inventory & Registry
@router.get("/agents", response_model=List[AgentApplication], tags=["Agent Registry"])
async def list_agents(
    tenant_id: Optional[str] = None,
    workspace_id: Optional[str] = None,
    category: Optional[AgentCategory] = None,
    state: Optional[AgentLifecycleState] = None,
):
    return orchestrator.registry.list_agents(tenant_id, workspace_id, category, state)


@router.post("/agents", response_model=AgentApplication, tags=["Agent Registry"])
async def register_agent(payload: Dict[str, Any] = Body(...)):
    cat_str = payload.get("category", "AUTOMATION")
    category = AgentCategory(cat_str) if cat_str in AgentCategory.__members__ else AgentCategory.AUTOMATION
    return orchestrator.registry.register_agent(
        tenant_id=payload["tenant_id"],
        organization_id=payload["organization_id"],
        workspace_id=payload["workspace_id"],
        name=payload["name"],
        slug=payload["slug"],
        category=category,
        owner_id=payload["owner_id"],
        owner_email=payload["owner_email"],
        description=payload.get("description", ""),
        tags=payload.get("tags", []),
    )


@router.get("/agents/{agent_id}", response_model=AgentApplication, tags=["Agent Registry"])
async def get_agent(agent_id: str):
    agent = orchestrator.registry.get_agent(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.put("/agents/{agent_id}/state", response_model=AgentApplication, tags=["Agent Registry"])
async def update_agent_lifecycle_state(agent_id: str, state: AgentLifecycleState = Query(...)):
    try:
        return orchestrator.registry.update_lifecycle_state(agent_id, state)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


# 3. Version Control (AVCS)
@router.get("/agents/{agent_id}/versions", response_model=List[AgentVersion], tags=["Agent Version Control"])
async def list_versions(agent_id: str):
    return orchestrator.versioning.list_versions(agent_id)


@router.post("/agents/{agent_id}/versions", response_model=AgentVersion, tags=["Agent Version Control"])
async def create_version(agent_id: str, payload: Dict[str, Any] = Body(...)):
    return orchestrator.versioning.create_version(
        agent_id=agent_id,
        version_tag=payload["version_tag"],
        system_prompt=payload["system_prompt"],
        tools=payload.get("tools", []),
        connectors=payload.get("connectors", []),
        model_family=payload.get("model_family", "gemini-pro"),
        changelog=payload.get("changelog", ""),
        accuracy_score=payload.get("accuracy_score", 0.95),
        cost_per_execution_usd=payload.get("cost_per_execution_usd", 0.004),
        parameters=payload.get("parameters", {}),
    )


@router.post("/agents/{agent_id}/rollback", response_model=AgentVersion, tags=["Agent Version Control"])
async def rollback_version(agent_id: str, target_version_tag: str = Query(...)):
    try:
        ver = orchestrator.versioning.rollback_to_version(agent_id, target_version_tag)
        orchestrator.registry.update_lifecycle_state(agent_id, AgentLifecycleState.DEPLOYED)
        return ver
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


# 4. Automated Testing
@router.post("/agents/{agent_id}/test", response_model=AgentTestResult, tags=["AI Quality & Security Testing"])
async def run_agent_tests(agent_id: str, version_tag: str = Query("1.0.0")):
    return orchestrator.testing.run_comprehensive_test_suite(agent_id, version_tag)


@router.get("/agents/{agent_id}/test-results", response_model=List[AgentTestResult], tags=["AI Quality & Security Testing"])
async def list_test_results(agent_id: str):
    return orchestrator.testing.list_test_results(agent_id)


# 5. Security Scanner
@router.post("/agents/{agent_id}/security-scan", response_model=AgentSecurityScan, tags=["Security Scanner"])
async def scan_agent_security(agent_id: str, payload: Dict[str, Any] = Body(...)):
    return orchestrator.security.scan_agent_version(
        agent_id=agent_id,
        version_tag=payload.get("version_tag", "1.0.0"),
        system_prompt=payload["system_prompt"],
        tools=payload.get("tools", []),
    )


@router.get("/agents/{agent_id}/security-scans", response_model=List[AgentSecurityScan], tags=["Security Scanner"])
async def list_security_scans(agent_id: str):
    return orchestrator.security.list_scans(agent_id)


# 6. Approval Workflow
@router.post("/agents/{agent_id}/approvals/submit", response_model=AgentApproval, tags=["Approval Workflow"])
async def submit_for_approval(agent_id: str, payload: Dict[str, Any] = Body(...)):
    return orchestrator.approvals.submit_for_approval(
        agent_id=agent_id,
        version_tag=payload.get("version_tag", "1.0.0"),
        submitter_id=payload["submitter_id"],
        submitter_email=payload["submitter_email"],
    )


@router.post("/agents/{agent_id}/approvals/review", response_model=AgentApproval, tags=["Approval Workflow"])
async def review_approval(agent_id: str, payload: Dict[str, Any] = Body(...)):
    stage_str = payload.get("stage", "FINAL_RELEASE")
    stage = ApprovalStage(stage_str) if stage_str in ApprovalStage.__members__ else ApprovalStage.FINAL_RELEASE
    dec_str = payload.get("decision", "APPROVED")
    decision = ApprovalDecision(dec_str) if dec_str in ApprovalDecision.__members__ else ApprovalDecision.APPROVED

    return orchestrator.approvals.review_and_decide(
        approval_id=payload["approval_id"],
        stage=stage,
        decision=decision,
        approver_id=payload["approver_id"],
        approver_email=payload["approver_email"],
        comments=payload.get("comments", ""),
    )


@router.get("/agents/{agent_id}/approvals", response_model=List[AgentApproval], tags=["Approval Workflow"])
async def list_approvals(agent_id: str):
    return orchestrator.approvals.list_approvals(agent_id)


# 7. Deployment Engine
@router.post("/agents/{agent_id}/deploy", response_model=AgentDeployment, tags=["Distributed Deployment"])
async def deploy_agent(agent_id: str, payload: Dict[str, Any] = Body(...)):
    env_str = payload.get("environment", "PRODUCTION")
    environment = DeploymentEnvironment(env_str) if env_str in DeploymentEnvironment.__members__ else DeploymentEnvironment.PRODUCTION
    strat_str = payload.get("strategy", "CANARY")
    strategy = DeploymentStrategy(strat_str) if strat_str in DeploymentStrategy.__members__ else DeploymentStrategy.CANARY

    dep = orchestrator.deployment.deploy_agent_version(
        agent_id=agent_id,
        version_tag=payload.get("version_tag", "1.0.0"),
        environment=environment,
        strategy=strategy,
        traffic_weight_pct=payload.get("traffic_weight_pct", 100),
        cluster_id=payload.get("cluster_id", "cluster_us_east_primary"),
    )
    orchestrator.registry.update_lifecycle_state(agent_id, AgentLifecycleState.DEPLOYED)
    return dep


@router.get("/agents/{agent_id}/deployments", response_model=List[AgentDeployment], tags=["Distributed Deployment"])
async def list_deployments(agent_id: str):
    return orchestrator.deployment.list_deployments(agent_id)


# 8. Dependency Graph
@router.get("/agents/{agent_id}/dependencies", response_model=List[AgentDependency], tags=["Dependency Graph"])
async def list_dependencies(agent_id: str):
    return orchestrator.dependencies.list_dependencies(agent_id)


@router.get("/agents/{agent_id}/dependencies/validate", tags=["Dependency Graph"])
async def validate_dependencies(agent_id: str):
    return orchestrator.dependencies.validate_dependency_graph(agent_id)


# 9. Starter Templates
@router.get("/templates", response_model=List[AgentTemplate], tags=["Starter Templates"])
async def list_templates(category: Optional[AgentCategory] = None):
    return orchestrator.templates.list_templates(category)


# 10. Marketplace
@router.get("/marketplace/listings", response_model=List[AgentMarketplaceListing], tags=["AI Marketplace"])
async def list_marketplace_listings(category: Optional[AgentCategory] = None):
    return orchestrator.marketplace.list_listings(category)


@router.post("/marketplace/publish", response_model=AgentMarketplaceListing, tags=["AI Marketplace"])
async def publish_to_marketplace(payload: Dict[str, Any] = Body(...)):
    cat_str = payload.get("category", "AUTOMATION")
    category = AgentCategory(cat_str) if cat_str in AgentCategory.__members__ else AgentCategory.AUTOMATION
    return orchestrator.marketplace.publish_agent(
        agent_id=payload["agent_id"],
        title=payload["title"],
        publisher_name=payload["publisher_name"],
        category=category,
        description=payload["description"],
        version=payload.get("version", "1.0.0"),
        price_monthly_usd=payload.get("price_monthly_usd", 0.0),
        tags=payload.get("tags", []),
    )


# 11. Analytics & ROI
@router.get("/analytics/{agent_id}", response_model=AgentAnalytics, tags=["Lifecycle Analytics"])
async def get_agent_analytics(agent_id: str):
    return orchestrator.analytics.get_agent_analytics(agent_id)


# 12. Retirement & Sunsetting
@router.post("/agents/{agent_id}/retire", response_model=AgentRetirementPlan, tags=["Retirement & Deprecation"])
async def schedule_agent_retirement(agent_id: str, payload: Dict[str, Any] = Body(...)):
    plan = orchestrator.retirement.schedule_retirement(
        agent_id=agent_id,
        deprecation_notice=payload.get("deprecation_notice", "Scheduled for retirement"),
        sunset_days=payload.get("sunset_days", 60),
        target_migration_agent_id=payload.get("target_migration_agent_id"),
    )
    orchestrator.registry.update_lifecycle_state(agent_id, AgentLifecycleState.DEPRECATED)
    return plan
